src/libtiled/tilelayer.cpp

```diff
@@ -114,7 +164,9 @@ void Tiled::TileLayer::setCell(int x, int y, const Cell &cell)
 {
     Q_ASSERT(contains(x, y));
 
-    Cell &existingCell = mGrid[x + y * mWidth];
+    Cell existingCell;
+
+    existingCell = chunk(x, y)->cellAt(x % CHUNK_SIZE, y % CHUNK_SIZE);
```

Comment:
This `x % CHUNK_SIZE` won't work for negative coordinates. Instead, with a chunk size of 16, we could do `x & 0xF`. We could call this `x & CHUNK_MASK` and define:
```c++
const int CHUNK_MASK = CHUNK_SIZE - 1;
```
Of course, this relies on the chunk size always being a power of 2.